franmolmedo/rambling-js-2/readme.md

```diff
@@ -0,0 +1,176 @@
+
+* Sin embargo, cuando tengamos un sólo argumento, los paréntesis son opcionales, tal y como hemos visto en el ejemplo previo.
+
+## Uso en casos reales
+
+Todo lo que se ha comentado está bien, pero realmente, y con los ejemplos vistos, tampoco parece un cambio demasiado relevante.
+Sin embargo, cuando nos vamos a casos un poco más realistas, es donde empezamos a observar su valía. El ejemplo más claro en el que es muy útil su uso es en los predicados que se utilizamos en las funciones para tratar con colecciones.
+Mostramos el típico ejemplo de un array de números, y queremos extraer aquellos cuyo valor sea mayor o igual que 5:
+
+```javascript
+const numbers = [10,2,3,4,5,8]
+
+const result = numbers.filter(function(number){
+    return number >= 5;
+});
+```
+
+En este caso, utilizando fat arrow functions, nuestro ejemplo quedaría de la siguiente forma:
+
+```javascript
+const numbers = [10,2,3,4,5,8]
+
+const result = numbers.filter(number => number >= 5);
+```
+
+La mejora queda clara en cuanto a legibilidad. Imaginemos su uso en callbacks, donde ayudaran, sino a reducir, sí a mejorar, el temido *callback hell* en javascript.
+
+## Comportamiento de `this`
+
+Todo lo comentado tiene su valor, pero donde realmente está el factor diferenciador de las fat arrow function respecto al empleo de las funciones tradicionales es en el cambio de comportamiento que ha sufrido `this`. Cualquier persona que haya trabajado con un código JavaScript relativamente complejo, saliendo de los ejemplos básicos, habrá obtenido comportamientos no esperables de `this`. ¿Cuántos hemos hecho uso de alguno de los siguientes *workarounds*?:
+* Guardarnos el contexto en una variable para poder utilizarlo en el callback. Todos hemos hecho alguna vez:
+```javascript
+function test(){
+    var self = this;
+    callOtherFunction(function(callbackResponse){
+        console.log(this);
+        console.log(self);
+    })
+}
+```
+
+* Utilizar el bind(this), a la hora de definir una función.
+
+Con el uso de las fat arrow function esto ya no será necesario, ya que, por defecto, `this` se va a referir al contexto donde la función haya sido declarada. Lo vemos con el ejemplo siguiente:
+
+```javascript
+const testObject = {
+    numbers: [10,20,30,40,50],
+    title: 'Double of',
+    double: function(){
+        return this.numbers.map(function(number){
+            return `${this.title} ${number} is ${2*number}`;
+        })
+    }
+}
+
+console.log(testObject.double());
+
+```
+
+Si ejecutamos el ejemplo obtenemos lo siguiente:
+
+```javascript
+["undefined 10 is 20", "undefined 20 is 40", "undefined 30 is 60", "undefined 40 is 80", "undefined 50 is 100"]
+```
+
+¿Qué está ocurriendo? Pues lo que todos sabíamos de antemano, el `this` dentro de la función que se ejecuta en el map no es el que cabría esperarse (el objeto testObject, sino que es el contexto global, donde, lógicamente, no está definido *title*). Esto lo resolveríamos en ES5 con una de las formas siguientes:
```

Comment:
SG: testObject y ES5 en cursiva. Yo pondría no está definido el this.title que queda mas claro.